File: DigitalImageProcessing/python/pointSet_minOutsourcing.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Funciton: 点集的最小外包：圆形、直立矩阵、旋转矩阵、三角形、凸多边形
@Python Version: 3.8
@Author: Wei Li
@Date: 2021-10-18
"""

import numpy as np
import cv2 as cv
            

# --------------------------
if __name__ == "__main__":
    print("-----------------------------------------------------")
    # -------- 点集的最小外包 --------
    # step 1, 最小外包旋转矩形
    # way 1 of point set store.
    pointSet = np.array([[1, 1], [5, 1], [1, 10], [5, 10], [2, 5]], np.int32)
    # way 2 of point set store.
    # pointSet = np.array([[[1, 1]], [[5, 1]], [[1, 10]], [[5, 10]], [[2, 5]]], np.int32)

    # compute the minimum outsourcing rotate rectangle of point set
    rotateRectangle = cv.minAreaRect(pointSet)
    # show infomation of rotate rectangle: [旋转矩阵的中心坐标，尺寸大小，旋转角度]
    print(f"The information of rotate rectangle: \n{rotateRectangle}")

    # 旋转矩形是通过中心点坐标、尺寸和旋转角度三个方面来定义的，
    # 通过这三个属性值就可以计算出旋转矩形的 4 个顶点，这样虽然简单，但是写起来比较复杂。
    # rotate rectangel
    vertices = cv.boxPoints(((200, 200), (90, 150), -60.0))
    print(f"The data type of rotate rectangle: {vertices.dtype}")
    print(f"The four points of rotate rectangle:\n {vertices}")
    # plot rotate rectangle with this four points
    img = np.zeros((400, 400), np.uint8)
    for idx in range(4):
        point_1 = vertices[idx, :]
        j = (idx+1) % 4
        point_2 = vertices[j, :]
        # plot line
        cv.line(img, (int(point_1[0]), int(point_1[1])), (int(point_2[0]), int(point_2[1])), 255, 2)
    cv.imshow("RotateRectangle", img)

    print("-----------------------------------------------------")
    # step 2, 最小外包圆形
    pointSetCircle = np.array([[1, 1], [5, 1], [1, 10], [5, 10], [2, 5]], np.int32)
    circle = cv.minEnclosingCircle(pointSetCircle)
    print(f"The infomation of this minimum outsouring circle: \n{circle}")
    
    print("-----------------------------------------------------")
    # step 3, 最小外包直立矩形
    pointSetRect = np.array([[[1, 1]], [[5, 10]], [[5, 1]], [[1, 10]], [[2, 5]]], np.float32)
    rightRect = cv.boundingRect(pointSetRect)
    # 返回结果是一个由 4 个元素组成的元组，
    # 前两个元素是直立矩形的一个顶点坐标，
    # 后两个元素是它的对角坐标
    print(f"The infomation of smallest outsourcing upright rectangle: \n{rightRect}")

    print("-----------------------------------------------------")
    # step 4, 最小凸包
    # 黑色画布 400X400
    convexImage = np.zeros((400, 400), np.uint8)
    # 随机生成 横纵 坐标均在 100-300 之间的坐标点
    num_randomPoint = 80
    pointSetConvex = np.random.randint(100, 300, (num_randomPoint, 2), np.int32)
    # visualization of this points
    for i in range(num_randomPoint):
        cv.circle(convexImage, (pointSetConvex[i, 0], pointSetConvex[i, 1]), 2, 255, 2)
    
    # compute the convex outsourcing of this poin set
    convex_hull = cv.convexHull(pointSetConvex)
    # 打印最外侧的点(连接起来即为凸多边形)
    print(f"The data type: {type(convex_hull)}")
    print(f"The data shape: {convex_hull.shape}")

    # 一次连接凸包里面的点
    k = convex_hull.shape[0]
    for i in range(k-1):
        cv.line(convexImage, (convex_hull[i, 0, 0], convex_hull[i, 0, 1]), (convex_hull[i+1, 0, 0], convex_hull[i+1, 0, 1]), 255, 2)
    # 首尾坐标相连
    cv.line(convexImage, (convex_hull[k-1, 0, 0], convex_hull[k-1, 0, 1]), (convex_hull[0, 0, 0], convex_hull[0, 0, 1]), 255, 2)
    cv.imshow("ConvexImage", convexImage)
    
    print("-----------------------------------------------------")
    # step 5, 最小外包三角形
    pointSetTriangle = np.array([[[1, 1]], [[5, 10]], [[5, 1]], [[1, 10]], [[2, 5]]], np.float32)
    area, triangle = cv.minEnclosingTriangle(pointSetTriangle)
    print(f"The area of triangle: {area}")
    print(f"The three point of triangle:\n {triangle}")
    print("-----------------------------------------------------")

    cv.waitKey()
    cv.destroyAllWindows()

File: DigitalImageProcessing/cpp/pointSet_minOutsourcing.cpp
/**
 * @File    : pointSet_minOutsourcing.cpp
 * @Brief   : 点集的最小外包：圆形、直立矩阵、旋转矩阵、三角形、凸多边形
 * @Author  : Wei Li
 * @Date    : 2021-10-18
*/

#include <iostream>
#include <vector>
#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>


// -------------------------------
int main(int argc, char **argv)
{
    std::cout << "--------------------------------------------------" << std::endl;
    // -------- 点集的最小外包 --------
    // step 1, 最小外包旋转矩形
    // point set
    // way 1 of point set store.
    // cv::Mat pointSet = (cv::Mat_<float>(5, 2) << 1, 1, 5, 1, 1, 10, 5, 10, 2, 5);

    // way 2 of point set store.
    // std::vector<cv::Point2f> pointSet;
    // pointSet.push_back(cv::Point2f(1, 1));
    // pointSet.push_back(cv::Point2f(5, 1));
    // pointSet.push_back(cv::Point2f(1, 10));
    // pointSet.push_back(cv::Point2f(5, 10));
    // pointSet.push_back(cv::Point2f(2, 5));

    // way 3 of point set store.
    cv::Mat pointSet = (cv::Mat_<cv::Vec2f>(5, 1) << cv::Vec2f(1, 1), cv::Vec2f(5, 1), cv::Vec2f(1, 10), cv::Vec2f(5, 10), cv::Vec2f(2, 5));

    // compute the minimum outsourcing rotate rectangle of point set
    cv::RotatedRect rotateRectangle = cv::minAreaRect(pointSet);
    // show infomation of rotate rectangle
    std::cout << "Angel of rotate rectangle: " << rotateRectangle.angle << std::endl;
    std::cout << "Center of rotate rectangle: " << rotateRectangle.center << std::endl;
    std::cout << "Size of rotate rectangle: " << rotateRectangle.size << std::endl;

    // 旋转矩形是通过中心点坐标、尺寸和旋转角度三个方面来定义的，
    // 通过这三个属性值就可以计算出旋转矩形的 4 个顶点，这样虽然简单，但是写起来比较复杂。
    // rotate rectangel
    cv::RotatedRect rotateRect (cv::Point2f(200, 200), cv::Point2f(90, 150), -60);
    //计算旋转矩形的 4 个顶点，存储为一个 4 行 2 列的单通道 float 类型的 Mat
    cv::Mat vertices;
    cv::boxPoints(rotateRect, vertices);
    std::cout << vertices << std::endl;
    // 在黑色发布上绘制该旋转矩形
    cv::Mat img = cv::Mat::zeros(cv::Size(400, 400), CV_8UC1);
    for (size_t i = 0; i < 4; ++i)
    {
        cv::Point point_1 = static_cast<cv::Point>(vertices.row(i));
        int j = (i + 1) % 4;
        cv::Point point_2 = static_cast<cv::Point>(vertices.row(j));
        cv::line(img, point_1, point_2, cv::Scalar(255), 3);
    }
    cv::imshow("RotateRectangel", img);

    std::cout << "--------------------------------------------------" << std::endl;
    // step 2, 最小外包圆形
    cv::Mat pointSetCircle = (cv::Mat_<float>(5, 2) << 1, 1, 5, 1, 1, 10, 5, 10, 2, 5);
    // compute the minimum outsourcing circle with this point set
    cv::Point2f center_circle;
    float radius_circle;
    cv::minEnclosingCircle(pointSetCircle, center_circle, radius_circle);
    // show infomation of the minimum outsourcing circle
    std::cout << "The center of circle: " << center_circle << std::endl;
    std::cout << "The radius of circle: " << radius_circle << std::endl;

    std::cout << "--------------------------------------------------" << std::endl;
    // step 3, 最小外包直立矩形
    std::vector<cv::Point2f> pointSetRect;
    pointSetRect.push_back(cv::Point2f(1, 1));
    pointSetRect.push_back(cv::Point2f(5, 1));
    pointSetRect.push_back(cv::Point2f(1, 10));
    pointSetRect.push_back(cv::Point2f(5, 10));
    pointSetRect.push_back(cv::Point2f(2, 5));

    // compute theThe smallest outsourcing upright rectangle with this point set
    cv::Rect uprightRectangle = cv::boundingRect(cv::Mat(pointSetRect));
    // show infomation of this upright rectangle
    std::cout << "The infomation of upright rectangle: " << uprightRectangle <<  std::endl;

    std::cout << "--------------------------------------------------" << std::endl;
    // step 4, 最小凸包
    // 5 行 2 列的单通道 Mat
    cv::Mat pointSetConvex = (cv::Mat_<float>(5, 2) << 1, 1, 5, 1, 1, 10, 5, 10, 2, 5);
    // compute the convex outsourcing of this poin set
    std::vector<cv::Point2f> hull;
    cv::convexHull(pointSetConvex, hull);
    // 打印最外侧的点(连接起来即为凸多边形)
    for (size_t i = 0; i < hull.size(); ++i)
    {
        std::cout << hull[i] << ", ";
    }
    std::cout << std::endl;

    std::cout << "--------------------------------------------------" << std::endl;
    // step 5, 最小外包三角形
    cv::Mat pointSetTriangle = (cv::Mat_<float>(5, 2) << 1, 1, 5, 1, 1, 10, 5, 10, 2, 5);
    pointSetTriangle = pointSetTriangle.reshape(2, 5); // 转换为双通道矩阵
    std::vector<cv::Point> triangle; // 存储三角形三个顶点
    double areaTriangle = cv::minEnclosingTriangle(pointSetTriangle, triangle);
    std::cout << "The infomation of triangle: " << std::endl;
    for (size_t i = 0; i < 3; ++i)
    {
        std::cout << triangle[i] << ", ";
    }
    std::cout << std::endl;
    std::cout << "The area of triangle: " << areaTriangle << std::endl;
    std::cout << "--------------------------------------------------" << std::endl;
    
    cv::waitKey();
    cv::destroyAllWindows();
    return 0;
}
